src/golang/cmd/server/handler/connect_integration.go

```diff
@@ -202,58 +207,241 @@ func (h *ConnectIntegrationHandler) Perform(ctx context.Context, interfaceArgs i
-				h.Database,
-			); err != nil {
-				log.Errorf("Unexpected error when setting the new storage layer: %v", err)
+			// Mark the migration explicitly as RUNNING.
+			runningAt := time.Now()
+			execState.Timestamps.RunningAt = &runningAt
+			err = h.updateStorageMigrationExecState(ctx, storageMigrationObj.ID, &execState)
+			if err != nil {
+				log.Errorf("Unexpected error when updating the storage migration entry to RUNNING: %v", err)
+				return
+			}
+
+			// TODO: REMOVE
+			if strings.HasPrefix(integrationObj.Name, "failing") {
+				err = errors.Newf("I REALLY DONT LIKE YOUR NAME %s SIR", integrationObj.Name)
+				return
+			}
+
+			// Actually perform the storage migration.
+			storageConfig, err := h.performStorageMigration(args.Service, args.Config, args.OrgID)
+			// We let the defer() handle the failure case appropriately.
+			if err != nil {
+				return
 			}
 
 			log.Info("Successfully migrated the storage layer!")
+			finishedAt := time.Now()
+			execState.Timestamps.FinishedAt = &finishedAt
+			execState.Status = shared.SucceededExecutionStatus
+
+			// The update of the storage config and storage migration entry should happen together.
+			// While we don't enforce this atomically, we can make the two update together to minimize the risk.
+			err = h.updateStorageMigrationExecState(ctx, storageMigrationObj.ID, &execState)
+			if err != nil {
+				log.Errorf("Unexpected error when updating the storage migration entry to SUCCESS: %v", err)
+				return
+			}
+
+			err = config.UpdateStorage(storageConfig)
+			if err != nil {
+				log.Errorf("Unexpected error when updating the global storage layer config: %v", err)
+			} else {
+				log.Info("Successfully updated the global storage layer config!")
+			}
 		}()
 	}
 
 	return emptyResp, http.StatusOK, nil
 }
 
-// ConnectIntegration connects a new integration specified by `args`. It returns a status code for the request
-// and an error, if any.
+// Also updates `current=True` if the execution state is marked as SUCCESS!
+func (h *ConnectIntegrationHandler) updateStorageMigrationExecState(
+	ctx context.Context,
+	storageMigrationID uuid.UUID,
+	execState *shared.ExecutionState,
+) error {
+	updates := map[string]interface{}{
+		models.StorageMigrationExecutionState: execState,
+	}
+
+	// This is updated to a transaction if we also need to mark an old entry as current=False.
+	db := h.Database
+	if execState.Status == shared.SucceededExecutionStatus {
+		updates[models.StorageMigrationCurrent] = true
+
+		// If there was a previous storage migration, update that entry to be `current=False`.
+		oldStorageMigrationObj, err := h.StorageMigrationRepo.Current(ctx, h.Database)
+		if err == nil {
+			// Updating the old storage migration entry must be done in the same transaction
+			// as the update of the new storage migration entry, so that there is at most one
+			// current=True entry in the storage_migration table.
+			txn, err := h.Database.BeginTx(context.Background())
```

Comment:
I would actually just create a `txn` at the start of this function. The read call of `StorageMigrationRepo.Current` should be inside of this transaction just from a database consistency standpoint. We wouldn't want to read from invalid state because a different operation modified the table between the read and when this transaction performs the update. 

Even if there is higher level concurrency control that prevents multiple storage migrations from happening at the same time, this would be good practice. It also lets us remove the semi-confusing logic of overwriting the `db` variable depending on if there was a previous storage migration or not.